ompi/mca/coll/base/coll_base_dynamic_file.c

```diff
@@ -0,0 +1,302 @@
+/*
+ * Copyright (c) 2020      Bull SAS. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "mpi.h"
+#include "ompi/mca/mca.h"
+#include "base.h"
+
+/* need to include our own topo prototypes so we can malloc data on the comm correctly */
+#include "coll_base_topo.h"
+
+/* need file reading function */
+#include "coll_base_util.h"
+
+/* also need the dynamic rule structures */
+#include "coll_base_dynamic_rules.h"
+
+/* and our own prototypes */
+#include "coll_base_dynamic_file.h"
+
+static int fileline=0; /* used for verbose error messages */
```

Comment:
This isn't thread safe.  I admittedly have only very, very quickly looked over this code -- is it possible for multiple MPI application threads to invoke these routines simultaneously and get mixedup in `fileline` values?

Reply:
The config file is read once during MPI_Init. As the MPI standard stands now, there is no need to protect this code.